ambry-store/src/main/java/com.github.ambry.store/StatsManager.java

```diff
@@ -0,0 +1,206 @@
+    publishPeriodInSecs = config.publishPeriodInSecs;
+    metrics = new StatsManagerMetrics(registry);
+    this.time = time;
+    scheduler = Utils.newScheduler(1, false);
+  }
+
+  /**
+   * Start the stats manager by scheduling the periodic task that collect, aggregate and publish stats.
+   */
+  public void start() {
+    // random initial delay between 1 to 10 minutes to offset nodes from collecting stats at the same time
+    statsAggregator = new StatsAggregator();
+    scheduler.scheduleAtFixedRate(statsAggregator, 0, publishPeriodInSecs, TimeUnit.SECONDS);
+  }
+
+  /**
+   * Stops the periodic task that is collecting, aggregating and publishing stats.
+   */
+  public void shutdown() throws InterruptedException {
+    if (statsAggregator != null) {
+      statsAggregator.cancel();
+    }
+    scheduler.shutdown();
+    if (!scheduler.awaitTermination(30, TimeUnit.SECONDS)) {
+      logger.error("Could not terminate aggregator tasks after StatsManager shutdown");
+    }
+  }
+
+  /**
+   * Publishes stats to a local file in JSON format.
+   * @param statsWrapper the {@link StatsWrapper} to be published
+   * @throws IOException
+   */
+  void publish(StatsWrapper statsWrapper) throws IOException {
+    File tempFile = new File(statsOutputFile.getAbsolutePath() + ".tmp");
+    if (tempFile.createNewFile()) {
+      OutputStream outputStream = new BufferedOutputStream(new FileOutputStream(tempFile));
+      DatumWriter<StatsWrapper> statsWrapperDatumWriter = new SpecificDatumWriter<StatsWrapper>(StatsWrapper.class);
+      JsonEncoder jsonEncoder = EncoderFactory.get().jsonEncoder(StatsWrapper.getClassSchema(), outputStream, true);
+      try {
+        statsWrapperDatumWriter.write(statsWrapper, jsonEncoder);
+        jsonEncoder.flush();
+        outputStream.flush();
+      } finally {
+        outputStream.close();
+      }
+      if (!tempFile.renameTo(statsOutputFile)) {
+        throw new IOException(
+            "Failed to rename " + tempFile.getAbsolutePath() + " to " + statsOutputFile.getAbsolutePath());
+      }
+    } else {
+      throw new IOException("Temporary file creation failed when publishing stats " + tempFile.getAbsolutePath());
+    }
+  }
+
+  /**
+   * Fetch and aggregate stats from a given {@link Store}
+   * @param aggregatedSnapshot the {@link StatsSnapshot} to hold the aggregated result
+   * @param partitionId specifies the {@link Store} to be fetched from
+   * @return true if stats were successfully fetched from the given {@link PartitionId}, false otherwise
+   */
+  boolean collectAndAggregate(StatsSnapshot aggregatedSnapshot, PartitionId partitionId) {
+    long fetchAndAggregatePerStoreStartTimeMs = time.milliseconds();
+    Store store = storageManager.getStore(partitionId);
+    if (store == null) {
+      return false;
+    } else {
+      StatsSnapshot statsSnapshot = null;
+      try {
+        statsSnapshot = store.getStoreStats().getStatsSnapshot();
+      } catch (StoreException e) {
```

Comment:
Ok looks like the caller doesn't but that is probably where you should handle all exceptions (instead of doing it in multiple places)